### ISSUE TITLE
Delete the last version of a schema

### ISSUE THREAD

**Issue by janvyhnanek** [Apicurio/apicurio-registry#2016]
As far as I know the current API and GUI do not allow user to delete only last version of a schema, the schema can be deleted only with all versions. We need to delete only the last version of a schema when someone registers by mistake a new version of schema which cannot be fixed by adding only compatible changes (mandatory backwards and forward compatibitlity is something we want and have set globally for all schemas). This situation occurs relatively often. 

Is it possible to delete the last version somehow nowadays or, is it possible to add such feature at least to apicurio API? Currently we use a workaround which is not very convenient - download all schema versions, delete schema and upload downloaded schema versions again without the last one. This workaround is also a bit dangerous for production use.

Thank you

**Comment by EricWittmann** [comment]
Originally our thought was to keep versions immutable.  That is why versions cannot be deleted, although they can be deprecated and disabled.  Have you tried disabling the offending version?

@Apicurio/developers any thoughts on this?  Should we change our position on allowing versions to be deleted?

**Comment by carlesarnal** [comment]
I don't think we want to make versions mutable again. That said, we can of course evolve how the states disabled or deprecated are being handled, AFAIK, even if you disable a version, if that version is the last one, it will be used for the compatibility checks.  So maybe we can consider looking for the latest *non-disabled* version?

**Comment by EricWittmann** [comment]
I think this is the right approach - allow disabling of versions (we do already) and eliminate them from consideration for things like compatibility checks.  And any other similar logic where we should be ignoring disabled versions.  It introduces some complexity, like what happens if you disable **every** version of an artifact.  We'll need to think through the edge cases a bit.

**Comment by janvyhnanek** [comment]
Sorry, I had probably made a mistake, we played a bit with settings of compatibility and I realized the reported problem is not an issue if backwards and forwards compatibility is switched on - deletion of the last version of a schema should not be necessary for compatibility checks. The reasoning is that starting from the second version of a schema you can only add optional elements with defaults so that the change is fully compatible and, as a result, this element can be always removed in a future version of a schema in a compatible way. If the wrong version of a schema is not deleted, there would be a version of schema in schema registry with irrelevant structure in it, but it should not bother anything during registration or runtime.

However, it might be an issue if only backwards or forwards compatibility are switched on separately, then the behaviour with checking only last ENABLED version looks reasonable to me.

**Comment by EricWittmann** [comment]
I think what we would need to do is just document the expected behavior when disabling a version.  Make a list of all the use-cases (e.g. when the version is the latest one, when it's not, when performing compatibility checking on a new version, etc etc).

Then we can identify where we have bugs (we may have many, as enable/disable is not a heavily used feature) and get them fixed.

I don't think it's fair to say that enable/disable does nothing.  But it is fair to say it's probably woefully incomplete and needs attention.  Let's give it some attention!  :)

**Comment by janvyhnanek** [comment]
I have seen that in the newer versions of Apicurio API there is now operation deleteArtifactVersion which does exactly what we wanted. It would be also useful to have this functionality in GUI, now it is still possible to delete only the whole schema with all its version (I tried version 2.5.7). Do you plan to add such feature to GUI?

thanks

**Comment by carlesarnal** [comment]
We will likely get there, but, as you can imagine, implementing such an operation in the UI is currently a low priority, especially given that this is already possible through the API.

**Comment by carlesarnal** [comment]
This was implemented in 3.0.
